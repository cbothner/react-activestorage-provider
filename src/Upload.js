/**
 * Perform the direct upload of a single file.
 *
 * @providesModule Upload
 * @flow
 */

import * as ActiveStorage from 'activestorage'

import type { ActiveStorageFileUpload, Origin } from './types'

const CONVENTIONAL_DIRECT_UPLOADS_PATH = '/rails/active_storage/direct_uploads'

type Options = {|
  origin?: Origin,
  directUploadsPath?: string,
  onBeforeBlobRequest?: ({
    id: string,
    file: File,
    xhr: XMLHttpRequest,
  }) => mixed,
  onBeforeStorageRequest?: ({
    id: string,
    file: File,
    xhr: XMLHttpRequest,
  }) => mixed,
  onChangeFile: ({ [string]: ActiveStorageFileUpload }) => mixed,
|}

class Upload {
  static defaultOptions = {
    origin: {},
    directUploadsPath: CONVENTIONAL_DIRECT_UPLOADS_PATH,
  }

  directUpload: ActiveStorage.DirectUpload
  options: { ...Options, ...typeof Upload.defaultOptions }

  get id(): string {
    return `${this.directUpload.id}`
  }

  get directUploadsUrl(): string {
    const {
      origin: { host, protocol, port },
      directUploadsPath,
    } = this.options

    if (host) {
      const builtProtocol = protocol
        ? `${protocol.split(':')[0]}://`
        : 'https://'
      const builtPort = port ? `:${port}` : ''
      return `${builtProtocol}${host}${builtPort}${directUploadsPath}`
    }
    return directUploadsPath
  }

  constructor(file: File, options: Options) {
    this.options = { ...Upload.defaultOptions, ...options }
    this.directUpload = new ActiveStorage.DirectUpload(
      file,
      this.directUploadsUrl,
      this
    )

    this.handleChangeFile({ state: 'waiting', id: this.id, file })
  }

  start(): Promise<string> {
    const promise = new Promise((resolve, reject) => {
      this.directUpload.create((error, attributes) => {
        if (error) reject(error)
        else resolve(attributes.signed_id)
      })
    })

    return promise.then(this.handleSuccess).catch(this.handleError)
  }

  /**
   * DirectUpload delegate protocol conformance
   */

  directUploadWillCreateBlobWithXHR(xhr: XMLHttpRequest) {
    this.options.onBeforeBlobRequest &&
      this.options.onBeforeBlobRequest({
        id: this.id,
        file: this.directUpload.file,
        xhr,
      })
  }

  directUploadWillStoreFileWithXHR(xhr: XMLHttpRequest) {
    this.options.onBeforeStorageRequest &&
      this.options.onBeforeStorageRequest({
        id: this.id,
        file: this.directUpload.file,
        xhr,
      })

    xhr.upload.addEventListener('progress', this.handleProgress)
  }

  /**
   * @private
   */

  handleChangeFile = (upload: ActiveStorageFileUpload) => {
    this.options.onChangeFile({ [this.id]: upload })
  }

  handleProgress = ({ loaded, total }: ProgressEvent) => {
    const progress = (loaded / total) * 100

    this.handleChangeFile({
      state: 'uploading',
      file: this.directUpload.file,
      id: this.id,
      progress,
    })
  }

  handleSuccess = (signedId: string) => {
    this.handleChangeFile({
      state: 'finished',
      id: this.id,
      file: this.directUpload.file,
    })
    return signedId
  }

  handleError = (error: string) => {
    this.handleChangeFile({
      state: 'error',
      id: this.id,
      file: this.directUpload.file,
      error,
    })
    throw error
  }
}

export default Upload
