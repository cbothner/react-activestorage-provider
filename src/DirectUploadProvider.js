/**
 * A component that creates an ActiveStorage::Blob in the Rails database,
 * uploads the files directly to the storage service and calls a render function
 * prop to allow the consumer to display the uploadâ€™s progress. On completion,
 * it calls back with the signed ids of the created blob objects.
 *
 * @providesModule DirectUploadProvider
 * @flow
 */

import * as React from 'react'

import Upload from './Upload'

import type {
  ActiveStorageFileUpload,
  Origin,
  RenderProps,
  CustomHeaders,
} from './types'

export type DelegatedProps = {|
  directUploadsPath?: string,
  multiple?: boolean,
  onBeforeBlobRequest?: ({
    id: string,
    file: File,
    xhr: XMLHttpRequest,
  }) => mixed,
  onBeforeStorageRequest?: ({
    id: string,
    file: File,
    xhr: XMLHttpRequest,
  }) => mixed,
  render: RenderProps => React.Node,
|}

type Props = {
  ...DelegatedProps,
  origin: Origin,
  onSuccess: (string[]) => mixed,
  headers?: CustomHeaders,
}

type State = {|
  uploading: boolean,
  fileUploads: { [string]: ActiveStorageFileUpload },
|}

class DirectUploadProvider extends React.Component<Props, State> {
  static defaultProps = {
    origin: {},
  }

  state = {
    uploading: false,
    fileUploads: {},
  }

  uploads: Upload[] = []

  render() {
    const { fileUploads } = this.state
    return this.props.render({
      handleChooseFiles: this.handleChooseFiles,
      handleBeginUpload: this.handleBeginUpload,
      handleUpload: this.handleUpload,
      ready: !this.state.uploading,
      uploads: Object.keys(fileUploads).map(key => fileUploads[key]),
    })
  }

  handleUpload = async (files: FileList | File[]) => {
    this.handleChooseFiles(files)
    return this.handleBeginUpload()
  }

  handleChooseFiles = (files: FileList | File[]) => {
    if (this.state.uploading) return

    this.setState({ fileUploads: {} })
    this.uploads = [...files].map(file => this._createUpload(file))
  }

  handleBeginUpload = async () => {
    if (this.state.uploading) return

    this.setState({ uploading: true })

    const signedIds = await Promise.all(
      this.uploads.map(upload => upload.start())
    )

    this.props.onSuccess(signedIds)
    this.uploads = []
    this.setState({ fileUploads: {}, uploading: false })
  }

  handleChangeFileUpload = (fileUpload: {
    [string]: ActiveStorageFileUpload,
  }) =>
    this.setState(({ fileUploads }) => ({
      fileUploads: { ...fileUploads, ...fileUpload },
    }))

  _createUpload(file: File) {
    const {
      directUploadsPath,
      headers,
      onBeforeBlobRequest,
      onBeforeStorageRequest,
      origin,
    } = this.props

    return new Upload(file, {
      directUploadsPath,
      headers,
      onBeforeBlobRequest,
      onBeforeStorageRequest,
      onChangeFile: this.handleChangeFileUpload,
      origin,
    })
  }
}

export default DirectUploadProvider
