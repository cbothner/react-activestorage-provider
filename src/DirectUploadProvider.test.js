/**
 * @noflow
 */

import React from 'react'
import renderer from 'react-test-renderer'
import DirectUploadProvider from './DirectUploadProvider'
import Upload from './Upload'

global.fetch = require('jest-fetch-mock')

jest.mock('Upload', () =>
  jest.fn((file, { onChangeFile }) => {
    onChangeFile({ [Date.now()]: { fileName: file.name } })

    return {
      start: jest.fn(() => {
        return Promise.resolve('signedId')
      }),
    }
  })
)

const onSuccess = jest.fn()

const file = new File([], 'file')

const uploadOptions = {
  directUploadsPath: 'direct_uploads',
  headers: { 'X-Custom': true },
  onBeforeBlobRequest: jest.fn(),
  onBeforeStorageRequest: jest.fn(),
  origin: {},
}

function renderComponent(props) {
  return renderer.create(
    <DirectUploadProvider
      {...uploadOptions}
      onSuccess={onSuccess}
      render={props => <div {...props} />}
      {...props}
    />
  )
}

beforeEach(() => {
  jest.clearAllMocks()
  fetch.resetMocks()
})

describe('DirectUploadProvider', () => {
  let component, tree

  beforeEach(() => {
    component = renderComponent()
    tree = component.toJSON()
  })

  it('updates the UI with upload progress', () => {
    expect(tree).toMatchSnapshot()

    tree.props.handleUpload([file])
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('creates and starts an upload when handleUpload is called', () => {
    tree.props.handleUpload([file])
    expect(Upload).toHaveBeenCalledWith(
      file,
      expect.objectContaining(uploadOptions)
    )
    expect(Upload.mock.results[0].value.start).toHaveBeenCalled()
  })

  it('creates an upload when handleChooseFiles is called', () => {
    tree.props.handleChooseFiles([file])
    expect(Upload).toHaveBeenCalledWith(
      file,
      expect.objectContaining(uploadOptions)
    )
    expect(Upload.mock.results[0].value.start).not.toHaveBeenCalled()
  })

  it('updates the file upload progress list when handleChooseFiles is called', () => {
    tree.props.handleChooseFiles([file])
    tree = component.toJSON()
    expect(tree.props.uploads).toHaveLength(1)

    tree.props.handleChooseFiles([file])
    tree = component.toJSON()
    expect(tree.props.uploads).toHaveLength(1)
  })

  it('starts the upload when handleBeginUpload is called', () => {
    tree.props.handleChooseFiles([file])

    tree.props.handleBeginUpload()
    expect(Upload.mock.results[0].value.start).toHaveBeenCalled()
  })

  it('calls onSuccess with [] if handleBeginUpload is called with no chosen files', async () => {
    await tree.props.handleBeginUpload()
    expect(onSuccess).toHaveBeenCalledWith([])
  })

  it('calls onSuccess prop when uploads are finished', async () => {
    await tree.props.handleUpload([file])
    expect(onSuccess).toHaveBeenCalledWith(['signedId'])
  })
})
