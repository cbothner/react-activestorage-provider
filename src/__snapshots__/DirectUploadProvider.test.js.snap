// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirectUploadProvider updates the UI with upload progress 1`] = `
<div
  handleBeginUpload={[Function]}
  handleChooseFiles={[Function]}
  handleUpload={[Function]}
  ready={true}
  uploads={Array []}
/>
`;

exports[`DirectUploadProvider updates the UI with upload progress 2`] = `
<div
  handleBeginUpload={[Function]}
  handleChooseFiles={[Function]}
  handleUpload={[Function]}
  ready={false}
  uploads={
    Array [
      Object {
        "fileName": "file",
      },
    ]
  }
/>
`;
